[18.0][MIG] sale_order_carrier_auto_assign: Migration to 18.0

lef-adhoc opened the pull request.

grindtildeath committed:
Create module sale_order_carrier_auto_assign

grindtildeath committed:
Use context key to avoid breaking other modules tests

oca-travis committed:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

Commit:
Compute shipping rate on carrier auto-assign

OCA-git-bot committed:
sale_order_carrier_auto_assign 13.0.1.1.0

danimv5 committed:
Added translation using Weblate (Spanish)

danimv5 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (1 of 1 strings)

Translation: sale-workflow-13.0/sale-workflow-13.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-13-0/sale-workflow-13-0-sale_order_carrier_auto_assign/es/

phuctranfxvn committed:
[MIG] sale_order_carrier_auto_assign: Migration to 14.0

oca-travis committed:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_order_carrier_auto_assign 14.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

Commit:
[MIG] sale_order_carrier_auto_assign: Migration to 15.0

Commit:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
sale_order_carrier_auto_assign 15.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

matiasperalta1 committed:
[MIG] sale_order_carrier_auto_assign: Migration to 16.0

santostelmo committed:
[IMP] sale_order_carrier_auto_assigni: add settings

Commit:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (1 of 1 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_carrier_auto_assign/it/

trisdoan committed:
[IMP] sale_order_carrier_auto_assign multi-company

Commit:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_carrier_auto_assign/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (7 of 7 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_carrier_auto_assign/it/

lef-adhoc committed:
[IMP] sale_order_carrier_auto_assign: pre-commit auto fixes

lef-adhoc committed:
[MIG] sale_order_carrier_auto_assign: Migration to 17.0

Commit:
[UPD] Update sale_order_carrier_auto_assign.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-17.0/sale-workflow-17.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-17-0/sale-workflow-17-0-sale_order_carrier_auto_assign/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-17.0/sale-workflow-17.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-17-0/sale-workflow-17-0-sale_order_carrier_auto_assign/

jjscarafia committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (6 of 6 strings)

Translation: sale-workflow-17.0/sale-workflow-17.0-sale_order_carrier_auto_assign
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-17-0/sale-workflow-17-0-sale_order_carrier_auto_assign/es/

lef-adhoc committed:
[IMP] sale_order_carrier_auto_assign: pre-commit execution

matiasperalta1 reviewed (approved):
LGTM

xaviedoanhduy commented on sale_order_carrier_auto_assign/models/sale_order.py:
```diff
@@ -0,0 +1,31 @@
+
+    def _add_delivery_carrier_on_confirmation(self):
+        """Automatically add delivery.carrier on sale order confirmation"""
+        for order in self:
```
I see `action_confirm` has iterated over the records in sale.order and called `_add_delivery_carrier_on_confirmation` function for each individual record. So why do we need to iterate over it again, you can consider it with function that directly handles that record and call `ensure_one` function will make the code clearer

xaviedoanhduy commented on sale_order_carrier_auto_assign/models/sale_order.py:
```diff
@@ -0,0 +1,31 @@
+                .with_context(**delivery_wiz_context)
+                .create({})
+            )
+            delivery_wiz._get_delivery_rate()
```
do we really need to call this function, I checked and found that it is mainly called in `onchange` functions and did not see it appear in any test in odoo codebase. I tried removing it and found that the module still works as expected and I guess that just calling `button_confirm` is enough, since in `button_confirm` there is a call to `set_delivery_line` function. I'm not sure if it's completely correct because I might miss something, do you think so?

lef-adhoc replied:
I deleted the line and I get an error in the tests and if I don't delete it I don't get an error. Can you check?

xaviedoanhduy replied:
thanks for taking the time, maybe it was worth appearing on that line.

xaviedoanhduy commented on sale_order_carrier_auto_assign/models/sale_order.py:
```diff
@@ -0,0 +1,31 @@
+
+    def _add_delivery_carrier_on_confirmation(self):
+        """Automatically add delivery.carrier on sale order confirmation"""
+        for order in self:
```
I see `action_confirm` has iterated over the records in `sale.order` and called `_add_delivery_carrier_on_confirmation` function for each single record. So why do we need to iterate over it again, you can consider it with function that directly handles that record and call `ensure_one` function will make the code clearer

lef-adhoc replied:
I did it another way, what do you think?

xaviedoanhduy replied:
it looks fine to me. thanks

xaviedoanhduy reviewed (commented):
hi @lef-adhoc, thanks for migration this module.
I tested the functional and it works fine.
btw, could you squash administrative commits.

xaviedoanhduy reviewed (approved):
LGTM

jbaudoux reviewed (changes requested):
You are missing this commit https://github.com/OCA/sale-workflow/commit/bfa00bb62fcdee6e8252983bb2396fb08fe4dbbb

and this change https://github.com/OCA/delivery-carrier/commit/bdff1b38c7664f96f79c9ed8fe71ad4c4fbe0ff4

and integrate those commits that were postponed for next version https://github.com/OCA/sale-workflow/pull/3081/commits

jbaudoux commented on sale_order_carrier_auto_assign/models/sale_order.py:
```diff
@@ -7,10 +7,9 @@ class SaleOrder(models.Model):
-            rec._add_delivery_carrier_on_confirmation()
+        self.filtered(
+            lambda so: so.company_id.carrier_auto_assign
+        )._add_delivery_carrier_on_confirmation()
```
This change should not be part of the migration commit

lef-adhoc replied:
I have already changed it to another commit

sebalix replied:
IMO it doesn't optimize previous code: now we are doing two loops, the `filtered` + the one in `_add_delivery_carrier_on_confirmation`.
With previous code we were looping only once on `self` (the loop in `_add_delivery_carrier_on_confirmation` can be ignored, it doesn't hurt with current recordset being of size 1).

So this commit can be removed.

lef-adhoc commented:
@jbaudoux This is the migration to 18 of the module that is already in 17 and has the same commits that are in 17 https://github.com/OCA/sale-workflow/commits/17.0/sale_order_carrier_auto_assign shouldn't you do the PRs first to 17 before including them in 18?

lef-adhoc committed:
[MIG] sale_order_carrier_auto_assign: Migration to 18.0

lef-adhoc committed:
[REF] sale: optimize `action_confirm` by filtering relevant records first

Replaced the loop with `filtered()` to improve code clarity and avoid
iterating over irrelevant records. The method `_add_delivery_carrier_on_confirmation`
is now only called on applicable orders directly.

sebalix commented:
/ocabot migration sale_order_carrier_auto_assign

victoralmau commented:
Please, cherry-pick https://github.com/OCA/sale-workflow/pull/3635 to commit history before migration commit.

simahawk commented:
@lef-adhoc are you going to finish this one?

jbaudoux commented:
Superseded by https://github.com/OCA/sale-workflow/pull/3709
